mailet/scala/src/main/scala/org/apache/james/transport/mailets/Throttler.scala

```diff
@@ -0,0 +1,143 @@
+  * <ul>
+  * <li><i>count</i> is expected to be an integer
+  * <li><i>period</i> is expected to be an finite duration greater than 1ms.
+  * It can be expressed using the scala duration DSL.
+  * </ul>
+  *</p>
+  * <pre><code>
+  * &lt;mailet match=&quot;All&quot; class=&quot;&lt;Throttler&gt;&quot;&gt;
+  * &lt;period&gt;1 minute&lt;/period&gt;
+  * &lt;count&gt;10&lt;/count&gt;
+  * &lt;/mailet&gt;
+  * </code></pre>
+  *
+  * <p>The scala Duration dsl expects a number (integer or fractional) followed by
+  * a unit. The unit can be one of
+  *<ul>
+  * <li> d|day|days, h|hour|hours
+  * <li> min|minute|minutes
+  * <li> s|second|seconds
+  * <li> ms|milli|millis|millisecond|milliseconds
+  * <li> µs|micro|micros|microsecond|microseconds
+  * <li> ns|nano|nanos|nanosecond|nanoseconds
+  *</ul>
+  * for exemple the following strings are valid durations:</p>
+  *<pre><code>
+  * 10 seconds
+  * 1 day
+  * 8h
+  * 100ns
+  * 80millis
+  *</code></pre>
+  *
+  */
+class Throttler(clock: Clock = Clock.systemUTC()) extends Mailet {
+  private var config: MailetConfig = _
+  private val safeRates: AtomicReference[Map[MailAddress, Seq[Instant]]] =
+    new AtomicReference(Map.empty.withDefaultValue(Seq.empty))
+
+  private var period: FiniteDuration = _
+  private var count: Int = _
+
+  override def init(config: MailetConfig): Unit = {
+    this.period = parsePeriod(config)
+    this.count = parseCount(config)
+    this.config = config
+  }
+
+  private def parseCount(config: MailetConfig): Int = Try(config.getInitParameter(Throttler.COUNT_PARAM).toInt)
+    .getOrElse(
+      throw new MailetException(
+        s"Missing or invalid configuration parameter ${Throttler.COUNT_PARAM} " +
+          s"for mailet ${config.getMailetName} ($getMailetInfo)"))
+
+  private def parsePeriod(config: MailetConfig): FiniteDuration = {
+    val duration = Try(Duration(config.getInitParameter(Throttler.PERIOD_PARAM)))
+      .getOrElse(
+        throw new MailetException(
+          s"Missing or invalid configuration parameter ${Throttler.PERIOD_PARAM} " +
+            s"for mailet ${config.getMailetName} ($getMailetInfo)"))
+    if (duration.isFinite()) {
+      if (duration.toMillis > 0) {
+        FiniteDuration(duration.toMillis, TimeUnit.MILLISECONDS)
+      } else {
+        throw new MailetException(
+          s"${Throttler.PERIOD_PARAM} must be a finite duration greater or equal to 1ms")
+      }
+    } else {
+      throw new MailetException(
+        s"${Throttler.PERIOD_PARAM} must be a finite duration greater or equal to 1ms")
+    }
+  }
+
+  override def service(mail: Mail): Unit = {
+    val sender = mail.getSender
+    val now = Instant.now(clock)
+    val limit = now.minusMillis(period.toMillis)
+    val updatedRates = safeRates.updateAndGet(rates => {
+      val instants = rates(sender)
+      rates + (sender -> (now +: instants.filter(_.isAfter(limit))))
+    })
+
+    if (updatedRates(sender).size > count) {
```

Comment:
Mmmh the action is hard-coded in the mailet thus it will not compose well.

Here are my thoughts:
 - What we are really dealing with is a **matcher** that then let the logic being expressed as configuration.

Instead of:

```
<mailet matcher="All" class="Throttler">
   <count >3600</count>
   <period>1 day</count>
</mailet>
```

I could write:

```
<mailet matcher="Throttler=count:3600&duration:1day" class="ToProcessor">
   <processor>handleThrotlingTheWayIWant</processor>
</mailet>
```

Here are some rationals:
 - Mail.ERROR state should IMO be used for unexpected error. If we know for sure where the error comes from we should try to introduce (or at least allow to introduce) a specific processor.
 - The caller should be able to specify what to do when the rate is exceeded. Should we log? Bounce? Ignore and silently drop the mail? Disallow relay but keep local delivery? Re-schedule local recipients for a later delivery?

As a general rule of thumb we should IMO try to separate `conditions` from `actions` when developing mailets/matchers.

Note that exposing the throttler as a **matcher** still allows to write a mailet relying on this matcher for convenience.

Reply:
Note that we may need a `SetErrorMessage` mailet